### ISSUE TITLE
Improve the exercise "linked-list"

### ISSUE THREAD

**Issue by siebenschlaefer** [exercism/c#466]
The [README.md](https://github.com/exercism/c/blob/master/exercises/linked-list/README.md) of the `linked-list` exercise says:

> You will write an implementation of a doubly linked list. Implement a Node to hold a value and pointers to the next and previous nodes. Then implement a List which holds references to the first and last node and offers an array-like interface for adding and removing items: [...]

That sounds like a common and well-known way to implement a doubly-linked list with a "control structure" that points to the first and last node and maybe even stores the length.

But the initial setup and the tests don't have that control structure `List`, just nodes. There, the list is a pointer to a pointer to a node. Where's the pointer to the first and especially the last node?
The [example implementation](https://github.com/exercism/c/blob/master/exercises/linked-list/src/example.c) doesn't have a dedicated last pointer, each `push` and `pop` has to traverse the whole list to get the last element. That linear runtime can be achieved with a singly-linked list, too.

IMHO there are two reasons to use doubly-linked list: You want to have access to the node structure to traverse the list in both directions, or you want performance characteristics that a singly-linked list cannot achieve. The tests dont's test the former or measure the latter, they even make implementing a control structure with a `first` and `last` node hard.

---

Why is `NULL` required to be a valid list? I understand `*list == NULL` is valid because there's no node. But requiring a list to be a
`struct list_item **` and then allowing the pointer to the pointer to be NULL feels like an invitation to a whole family of bugs.

---

From the [README.md](https://github.com/exercism/c/blob/master/exercises/linked-list/README.md):

> To keep your implementation simple, the tests will not cover error conditions. Specifically: `pop` or `shift` will never be called on an empty list.

IMHO the two tests `test_push_with_invalid_list` and `test_unshift_with_invalid_list` violate the spirit of that "no error
conditions".

---

The naming of the function is slightly inconsistent.

`new_list` and `delete_list` have `_list` at the end, `is_list_empty` has it in the middle, while the other functions don't mention `list`.  
*Personally* I'm a fan of prefixes, e.g. `list_create`, `list_destroy`, `list_is_empty`, `list_push`, etc.  
The exercise `react` uses a "verb_thing" scheme, e.g. `create_list`, `destroy_list`, `push_value` (and maybe `is_list_empty`?)

I'll use a `list_` prefix below, but I'm open to alternatives.

---

I would like to propose a different setup with a dedicated control structure `list`.

```c
// constructs a new list of items
struct list *list_create(void);

// checks if the list is empty
bool list_is_empty(struct list *list);

// inserts item at back of list
bool list_push(struct list *list, ll_data_t item_data);

// removes item from back of list
ll_data_t list_pop(struct list *list);

// removes item from front of list
ll_data_t list_shift(struct list *list);

// inserts item at front of list
bool list_unshift(struct list *list, ll_data_t item_data);

// destroy the entire list
// list will be a dangling pointer after calling this method on it
void list_destroy(struct list *list);
```

The `struct list` and the implicit `struct list_node` could be visible and explicit in the `.h` file:
```c
typedef int ll_data_t;
struct list_node
{
    struct list_node *prev, *next;
    ll_data_t;
};
struct list
{
    struct list_node *first, *last;
};
```
or they could be forward declared and hidden in the `.c` file:
```c
typedef int ll_data_t;
struct list;
```

This proposal would match the description in the `README.md` better (and it would allow an O(1) implementation of `push` and `pop`.)

**Comment by elyashiv** [comment]
Seems like valid point -

I wondered if it worth changing the exercise to something that looks more like windows `LIST_ENTRY` - keep just the next and prev pointers in the struct, and embed it in another struct for data.

**Comment by ryanplusplus** [comment]
> Seems like valid points.
> 
> I wondered if it worth changing the exercise to something that looks more like windows `LIST_ENTRY` - keep just the next and prev pointers in the struct, and embed it in another struct for data.

Do you mean an [intrusive list](https://www.gamasutra.com/view/news/128568/InDepth_Intrusive_Lists.php)?

**Comment by ryanplusplus** [comment]
@siebenschlaefer I think you're dead on; this exercise doesn't match the intent or the description. If you're willing to submit a PR to fix this I'd love to have it.

**Comment by siebenschlaefer** [comment]
@ryanplusplus Sure, I can do that. Do you prefer `struct list` and `struct list_node` to be defined in the `.h` file or just forward declared and left to the students?

**Comment by ryanplusplus** [comment]
> @ryanplusplus Sure, I can do that. Do you prefer `struct list` and `struct list_node` to be defined in the `.h` file or just forward declared and left to the students?

Given how easy it is to misunderstand the requirements I think having the structs actually be defined will be helpful.

**Comment by elyashiv** [comment]
You can define them in the c file, allowing them to be opaque. I think this is a better approach, especially if the test don't test the actual structure.

Also, this will give a students a nudge to learn about forward declaration ;)

**Comment by wolf99** [comment]
This was fixed in #467
